contributor-docs/adrs/adr-001-custom-elements.md

```diff
@@ -0,0 +1,115 @@
+
+```jsx
+export const ClipboardCopy = () => {
+ return <clipboard-copy for="span">Copy button</clipboard-copy>;
+};
+
+// passing a custom element to styled components results in no styles being applied
+const StyledClipboardCopy = styled(ClipboardCopy)`
+ background-color: red;
+`;
+
+const App = () => {
+ return (
+   <ThemeProvider theme={theme}>
+     <BaseStyles className="App">
+       <StyledClipboardCopy for=”span” />
+       <span id="span">Content from span tag.</span>
+     </BaseStyles>
+   </ThemeProvider>
+ );
+};
+
+```
+
+
+
+View in CodeSandbox: https://codesandbox.io/s/demo-styling-custom-element-g973d?file=/src/index.tsx:285-599
+
+Instead you would have to do something like this:
+
+```jsx
+const ClipboardCopy = () => {
+  return <clipboard-copy for="span">Copybutton</clipboard-copy>;
+};
+
+const StyledClipboardCopyWrapper = styled.span`
+  background-color: red;
+`;
+
+const App = () => {
+  return (
+   <>
+    <StyledClipboardCopyWrapper>
+     <ClipboardCopy />
+    </StyledClipboardCopyWrapper>
+    <span id="span">Content from span tag.</span>
+  </>
+ );
+};
+```
+
+
+
+View in CodeSandbox: https://codesandbox.io/s/demo-styling-custom-element-2-xt9g9?file=/src/index.tsx
+
+This also means we would need to be careful to never export a component with a custom element as the top level node.
+
+### Incompatibility with some React tools
+Some of our GitHub custom elements such as details-dialog and details-menu make assumptions about the DOM tree. For example, details-dialog expects a `details` element to wrap the custom element and uses this assumption[^2] to determine whether or not clicks are happening inside or outside of the dialog and closes the dialog if the click happened outside of the dialog. This makes sense in most cases and is a nice way of enforcing proper usage of the details element, but breaks down when used with [React Portals](https://reactjs.org/docs/portals.html) which are often used to ensure menus are displayed correctly in cases where a parent has an overflow: hidden applied to it, or incompatible z-index.
+
+### Extensibility
+Building behaviors in React Hooks gives us the ability to provide things like state and state change hooks to the consumer of the component. This allows the user to build on additional behaviors to the component based on the state or other variables provided to the component consumer. Doing the same with custom elements would require listening to events on the document[^3] and reacting to them. This is certainly do-able, but goes against some of the foundational principles of React (reacting to changes in the DOM vs changes in React state).
+
+### Organizational Overhead
+- GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.
```

Comment:
What additional maintenance overhead does this create?

Reply:
1. You'd need to `npm link` while developing if you want to test changes out with the presentational components themselves instead of making changes and seeing updates instantly. `npm link` usually doesn't work well with hot module reloading either.
2. You'd need to draft & publish releases to both libraries every time you want to update the behavior
3. If the behaviors are shared between .com and PRC you'd need to do careful testing in both environments to make sure that changes don't create any regressions. That greatly widens the context that engineers need to keep in mind every time a change is made.

Reply:
Let's enumerate these points in the ADR:

```suggestion
- GitHub’s custom elements are all managed in different repos which introduces more maintenance overhead.

  - You'd need to npm link while developing if you want to test changes out with the presentational components themselves instead of making changes and seeing updates instantly. npm link usually doesn't work well with hot module reloading either.
  - You'd need to draft & publish releases to both libraries every time you want to update the behavior
  - If the behaviors are shared between .com and PRC you'd need to do careful testing in both environments to make sure that changes don't create any regressions. That greatly widens the context that engineers need to keep in mind every time a change is made.
```